File: backend/user-service/services/member_service.go
package services

import (
	"context"
	"user-service/models"
	"user-service/models/dto"
	"user-service/repositories"

	"github.com/google/uuid"
)

type IMemberService interface {
	GetMemberProfile(ctx context.Context, userID uuid.UUID) (*dto.MemberProfileResponse, error)
	GetRawMemberProfile(ctx context.Context, userID uuid.UUID) (*models.MemberProfile, error)
	CreateMemberProfile(ctx context.Context, userID uuid.UUID) (*models.MemberProfile, error)
	UpdateMemberProfile(ctx context.Context, profile *models.MemberProfile) error
	DeleteMemberProfile(ctx context.Context, userID uuid.UUID) error
	GetMembersWithPagination(ctx context.Context, page, limit int, searchQuery *string) (*dto.MemberListResponse, error)
}

type MemberService struct {
	repo           repositories.IMemberRepository
	roleRepo       repositories.IRoleRepository
	entitlementSvc IEntitlementService
}

func NewMemberService(repo repositories.IMemberRepository, roleRepo repositories.IRoleRepository, entitlementSvc IEntitlementService) IMemberService {
	return &MemberService{
		repo:           repo,
		roleRepo:       roleRepo,
		entitlementSvc: entitlementSvc,
	}
}

// GetMemberProfile retrieves a member profile by user ID and enriches it with entitlements
func (s *MemberService) GetMemberProfile(ctx context.Context, userID uuid.UUID) (*dto.MemberProfileResponse, error) {
	profile, err := s.repo.FindByUserID(ctx, userID)
	if err != nil {
		return nil, err
	}

	// Fetch entitlements for this user
	entitlementMap, err := s.entitlementSvc.FindSessionsStatsByMemberIDs(ctx, []uuid.UUID{userID})
	var entitlements []models.Entitlement
	if err == nil {
		entitlements = entitlementMap[userID]
	}
	if entitlements == nil {
		entitlements = []models.Entitlement{}
	}

	// Compute total remaining sessions across all entitlements
	totalAvailable := 0
	for _, e := range entitlements {
		totalAvailable += e.Remaining
	}

	return &dto.MemberProfileResponse{
		UserID:                 profile.UserID,
		SessionsCompleted:      profile.SessionsCompleted,
		TrainingTime:           profile.TrainingTime,
		AverageRating:          profile.AverageRating,
		TotalAvailableSessions: totalAvailable,
		IdentityFullname:       profile.IdentityFullname,
		Entitlements:           entitlements,
	}, nil
}

// GetRawMemberProfile retrieves the raw model (used internally for updates)
func (s *MemberService) GetRawMemberProfile(ctx context.Context, userID uuid.UUID) (*models.MemberProfile, error) {
	return s.repo.FindByUserID(ctx, userID)
}

// CreateMemberProfile creates a new member profile for a user
func (s *MemberService) CreateMemberProfile(ctx context.Context, userID uuid.UUID) (*models.MemberProfile, error) {
	profile := &models.MemberProfile{
		UserID:            userID,
		SessionsCompleted: 0,
		TrainingTime:      0,
		AverageRating:     0,
	}
	if err := s.repo.Create(ctx, profile); err != nil {
		return nil, err
	}
	return profile, nil
}

// UpdateMemberProfile updates a member profile
func (s *MemberService) UpdateMemberProfile(ctx context.Context, profile *models.MemberProfile) error {
	return s.repo.Update(ctx, profile)
}

// DeleteMemberProfile deletes a member profile by user ID
func (s *MemberService) DeleteMemberProfile(ctx context.Context, userID uuid.UUID) error {
	return s.repo.Delete(ctx, userID)
}

// GetMembersWithPagination returns paginated list of members with their profiles
// Calculates totalAvailableSessions dynamically from active entitlements
func (s *MemberService) GetMembersWithPagination(ctx context.Context, page, limit int, searchQuery *string) (*dto.MemberListResponse, error) {
	// Find the "member" role
	roleModel, err := s.roleRepo.FindRoleByName(ctx, "member")
	if err != nil {
		return nil, err
	}

	// Get total count
	total, err := s.repo.CountByRoleID(ctx, roleModel.ID)
	if err != nil {
		return nil, err
	}

	// Calculate pagination offset
	offset := (page - 1) * limit

	// Determine search query - use empty string if nil
	query := ""
	if searchQuery != nil {
		query = *searchQuery
	}

	// Get paginated users
	users, err := s.repo.FindByRoleIDWithPagination(ctx, roleModel.ID, offset, limit, query)
	if err != nil {
		return nil, err
	}

	// Collect member IDs for bulk entitlements query
	memberIDs := make([]uuid.UUID, 0, len(users))
	for _, user := range users {
		if user.MemberProfile != nil {
			memberIDs = append(memberIDs, user.ID)
		}
	}

	// Get active entitlements for all members
	entitlements, err := s.entitlementSvc.FindSessionsStatsByMemberIDs(ctx, memberIDs)
	if err != nil {
		// Log but don't fail - just set available sessions to 0
		entitlements = make(map[uuid.UUID][]models.Entitlement)
	}

	// Convert to response DTOs
	data := make([]dto.UserWithProfileResponse, len(users))
	for i, user := range users {
		var memberProfile *dto.MemberProfileResponse
		if user.MemberProfile != nil {
			// Calculate total remaining sessions from active entitlements
			totalRemaining := 0
			if ents, ok := entitlements[user.ID]; ok {
				for _, ent := range ents {
					totalRemaining += ent.Remaining
				}
			}

			memberProfile = &dto.MemberProfileResponse{
				UserID:                 user.MemberProfile.UserID,
				SessionsCompleted:      user.MemberProfile.SessionsCompleted,
				TrainingTime:           user.MemberProfile.TrainingTime,
				AverageRating:          user.MemberProfile.AverageRating,
				TotalAvailableSessions: totalRemaining,
				IdentityFullname: 		user.MemberProfile.IdentityFullname,
				Entitlements:           entitlements[user.ID],
			}
		}

		data[i] = dto.UserWithProfileResponse{
			GetUserResponse: dto.GetUserResponse{
				UserID:      user.ID,
				Email:       user.EmailAddress,
				Username:    user.Username,
				FirstName:   user.FirstName,
				LastName:    user.LastName,
				PhoneNumber: user.PhoneNumber,
				Image:       user.Image,
				DateOfBirth: user.DateOfBirth,
				Address:     user.Address,
				RoleID:      user.RoleID,
				Role: dto.RoleResponse{
					ID:   user.RoleID,
					Name: user.Role.Name,
				},
			},
			MemberProfile: memberProfile,
		}
	}

	return &dto.MemberListResponse{
		Data:       data,
		Pagination: dto.NewPaginationMeta(total, page, limit),
	}, nil
}

File: app/services/memberService.ts
import type { ApiResponse, PaginatedResponse } from "~/composables/useApiClients";

// ========== Entitlement ==========

export interface Entitlement {
  id: string;
  memberId: string;
  bookingId: string;
  packageId: string;
  packageName: string;
  isNightSession: boolean;
  isWeekendSession: boolean;
  totalSessions: number;
  remaining: number;
  usedSessions: number;
  startDate: string;
  endDate: string | null;
  status: "active" | "completed" | "expired" | "cancelled";
  createdAt: string;
  updatedAt: string;
}

// ========== Member Profile ==========

export interface MemberProfile {
  userId: string;
  sessionsCompleted: number;
  trainingTime: number; // in minutes
  averageRating: number;
  totalAvailableSessions: number;
  entitlements: Entitlement[];
  identityFullname: string;
}

export interface UpdateMemberProfileData {
  identityFullname?: string;
}

// ========== Member List ==========

export interface MemberListItem {
  userId: string;
  email: string;
  username: string;
  firstName: string;
  lastName: string;
  phoneNumber: string;
  image: string;
  dateOfBirth: string;
  address: string;
  roleId: number;
  role: {
    id: number;
    name: string;
  };
  memberProfile?: Partial<MemberProfile>;
  createdAt: string;
}

export interface MemberListResponse {
  members: MemberListItem[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// ========== Member Service ==========

export const memberService = {
  /**
   * Get member profile by user ID
   * GET /members/{userId}/profile
   */
  async getMemberProfile(userId: string): Promise<MemberProfile | null> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<MemberProfile>>(
        `/members/${userId}/profile`,
        { method: "GET" },
      );
      return extractData(response);
    } catch {
      return null;
    }
  },

  /**
   * Get current user's member profile
   * GET /members/{userId}/profile
   */
  async getMyProfile(accessToken: string): Promise<MemberProfile | null> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<MemberProfile>>(
        "/members/me/profile",
        { method: "GET" },
      );
      return extractData(response);
    } catch {
      return null;
    }
  },

  /**
   * Update member profile
   * PUT /members/{userId}/profile
   */
  async updateMemberProfile(
    userId: string,
    data: UpdateMemberProfileData,
  ): Promise<MemberProfile | null> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<MemberProfile>>(
        `/members/${userId}/profile`,
        {
          method: "PUT",
          body: data,
        },
      );
      return extractData(response);
    } catch {
      return null;
    }
  },

  /**
   * Update user details
   * PUT /users/{id}
   */
  async updateUser(
    userId: string,
    data: {
      firstName: string;
      lastName: string;
      phoneNumber?: string;
      address?: string;
      dateOfBirth?: string;
    },
  ): Promise<any> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<any>>(
        `/users/${userId}`,
        {
          method: "PUT",
          body: data,
        },
      );
      return extractData(response);
    } catch {
      return null;
    }
  },

  /**
   * Get all members with pagination (admin)
   * GET /members
   */
  async getAllMembers(params: {
    page?: number;
    limit?: number;
    search?: string;
  } = {}): Promise<MemberListResponse> {
    const { user, extractPaginatedData } = useApiClients();

    const queryParams = new URLSearchParams();
    if (params.page) queryParams.set("page", String(params.page));
    if (params.limit) queryParams.set("limit", String(params.limit));
    if (params.search) queryParams.set("search", params.search);

    const queryString = queryParams.toString();
    const url = `/members${queryString ? `?${queryString}` : ""}`;

    try {
      const response = await user<PaginatedResponse<MemberListItem>>(url, {
        method: "GET",
      });
      const { data, pagination } = extractPaginatedData(response);
      return { members: data, pagination };
    } catch {
      return { members: [], pagination: { page: 1, limit: 10, total: 0, totalPages: 0 } };
    }
  },

  /**
   * Search members with pagination
   * GET /members/search
   */
  async searchMembers(params: {
    page?: number;
    limit?: number;
    search: string;
  }): Promise<MemberListResponse> {
    const { user, extractPaginatedData } = useApiClients();

    const queryParams = new URLSearchParams();
    if (params.page) queryParams.set("page", String(params.page));
    if (params.limit) queryParams.set("limit", String(params.limit));
    queryParams.set("search", params.search);

    const url = `/members/search?${queryParams.toString()}`;

    try {
      const response = await user<PaginatedResponse<MemberListItem>>(url, {
        method: "GET",
      });
      const { data, pagination } = extractPaginatedData(response);
      return { members: data, pagination };
    } catch {
      return { members: [], pagination: { page: 1, limit: 10, total: 0, totalPages: 0 } };
    }
  },

  /**
   * Get recent registrations (admin)
   * GET /members/recent
   */
  async getRecentRegistrations(params: {
    limit?: number;
    fromDate?: string;
    toDate?: string;
  } = {}): Promise<MemberListItem[]> {
    const { user, extractData } = useApiClients();

    const queryParams = new URLSearchParams();
    if (params.limit) queryParams.set("limit", String(params.limit));
    if (params.fromDate) queryParams.set("fromDate", params.fromDate);
    if (params.toDate) queryParams.set("toDate", params.toDate);

    const queryString = queryParams.toString();
    const url = `/members/recent${queryString ? `?${queryString}` : ""}`;

    try {
      const response = await user<ApiResponse<MemberListItem[]>>(url, {
        method: "GET",
      });
      return extractData(response);
    } catch {
      return [];
    }
  },

  /**
   * Get member entitlements (sessions/packages)
   * GET /members/{userId}/entitlements
   */
  async getMemberEntitlements(userId: string): Promise<Entitlement[]> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<Entitlement[]>>(
        `/members/${userId}/entitlements`,
        { method: "GET" },
      );
      return extractData(response);
    } catch {
      return [];
    }
  },

  /**
   * Get current user's entitlements
   */
  async getMyEntitlements(): Promise<Entitlement[]> {
    const { user, extractData } = useApiClients();

    try {
      const response = await user<ApiResponse<Entitlement[]>>(
        "/members/me/entitlements",
        { method: "GET" },
      );
      return extractData(response);
    } catch {
      return [];
    }
  },
};
